### PR TITLE
[ONNX] Reshape op

### PR DESCRIPTION
**Pull request by CircleSpin**
This PR adds the new allowzero parameter described in onnx opset 14 for reshape: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Reshape

allowzero set to true means that a zero in the shape tensor will be treated as a true zero value (aka an empty tensor), as opposed to it signifying to copy the value at the equivalent position in the data tensor. 

It's been a journey. Thanks @jwfromm for co-venturing :ship: :world_map: 

Thoughts? @MargaretQian @anwang2009 @sfvaroglu

### CODE REVIEW

File: tests/python/frontend/onnx/test_forward.py [hunk 1]
```diff
@@ -5077,7 +5077,7 @@ def verify_eyelike(indata):
     "test_reduce_sum_keepdims_random",
     "test_reduce_sum_negative_axes_keepdims_example",
     "test_reduce_sum_negative_axes_keepdims_random",
-    "test_reshape_allowzero_reordered",
+    # "test_reshape_allowzero_reordered",
```
Comment (jwfromm) [inline tests/python/frontend/onnx/test_forward.py:5080]: I think we probably want to just delete this line instead of commenting it out.

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -634,11 +636,17 @@ Array<IndexExpr> InferNewShape(const Array<IndexExpr>& data_shape, const Attrs&
       oshape.push_back(newshape[i]);
       ++src_idx;
     } else if (svalue == 0) {
-      // keep same
-      ICHECK_LT(src_idx, ishape.size());
-      used_input_dims.insert(src_idx);
-      used_output_dims.insert(oshape.size());
-      oshape.push_back(ishape[src_idx++]);
+      if (allowzero) {
+        // keep same as > 0
```
Comment (jwfromm) [inline src/relay/op/tensor/transform.cc:640]: This comment could be clearer, especially since "keep same" was used to describe the allowzero = false condition :)

File: src/relay/op/make_op.h [hunk 3]
```diff
@@ -67,7 +67,7 @@ Expr MakeReduce(Expr data, Array<Integer> axis, bool keepdims, bool exclude, Str
 
 Expr MakeRepeat(Expr data, int repeats, int axis);
 
-Expr MakeReshape(Expr data, Array<Integer> newshape);
+Expr MakeReshape(Expr data, Array<Integer> newshape, int allowzero = 0);
```
Comment (jwfromm) [inline src/relay/op/make_op.h:70]: You use `bool` for allowzero in the dynamic `MakeReshape`. Let's also use it here (and in transform.cc) instead of `int` for consistency.

File: include/tvm/topi/transform.h [hunk 4]
```diff
@@ -321,6 +321,11 @@ inline Tensor reshape(const Tensor& x, Array<PrimExpr> newshape, std::string nam
   auto x_shape = x->shape;
   Array<PrimExpr> target_shape;
 
+  // if newshape is an empty tensor make it [0] instead.
+  if (newshape.empty()) {
+    newshape.push_back(0);
+  }
```
Comment (jwfromm) [inline include/tvm/topi/transform.h:327]: I think we decided this snippet is actually not needed. Can you remove it and confirm everything still works?

File: include/tvm/topi/detail/ravel_unravel.h [hunk 5]
```diff
@@ -70,8 +70,12 @@ inline Array<PrimExpr> UnravelIndex(PrimExpr idx, Array<PrimExpr> shape) {
   std::vector<PrimExpr> indices;
 
   for (int i = static_cast<int>(shape.size()) - 1; i >= 0; --i) {
-    indices.push_back(indexmod(idx, shape[i]));
-    idx = indexdiv(idx, shape[i]);
+    if (shape[i].as<IntImmNode>()->value == 0) {
```
Comment (jwfromm) [inline include/tvm/topi/detail/ravel_unravel.h:73]: I just wanted to call out this bit for other reviewers. To support empty tensor reshapes we have to insert a check in the unravel inner loop. This could affect performance but is the only way to enable the functionality. Does anyone have any strong thoughts on this approach?

**Review by jwfromm — CHANGES_REQUESTED** [review]

File: src/relay/op/make_op.h (hunk 3 above)
Comment (CircleSpin) [reply src/relay/op/make_op.h:70]: int it always best to be consistent? so booliant (brilliant) :P

File: src/relay/op/tensor/transform.cc (hunk 2 above)
Comment (CircleSpin) [reply src/relay/op/tensor/transform.cc]: Gave it a quick rewrite, thoughts?

File: tests/python/frontend/onnx/test_forward.py (hunk 1 above)
Comment (CircleSpin) [reply tests/python/frontend/onnx/test_forward.py]: bye bye test :wave:

File: include/tvm/topi/transform.h (hunk 4 above)
Comment (CircleSpin) [reply include/tvm/topi/transform.h]: :+1:

**Comment by sfvaroglu** [comment]
@AndrewZhaoLuo mind taking a look?

**Comment by AndrewZhaoLuo** [comment]
I'll take a look tomorrow. This looks potentially like one that will require attention

File: python/tvm/relay/frontend/onnx.py [hunk 6]
```diff
@@ -1203,11 +1203,14 @@ def _impl_v1(cls, inputs, attr, params):
 
     @classmethod
     def _impl_v5(cls, inputs, attr, params):
+        allowzero = False
```
Comment (AndrewZhaoLuo) [inline python/tvm/relay/frontend/onnx.py:1206]: suggest attr.get('allowzero', False)

File: python/tvm/relay/op/dyn/_transform.py [hunk 7]
```diff
@@ -45,12 +45,15 @@ def _reshape_shape_func_input_data(data_shape, newshape, ndim):
     for i in const_range(len(newshape)):
         if skip > 0:
             skip -= 1
-        elif newshape[i] > 0:
+        elif newshape[i] >= 0:
             out[dst_idx] = int64(newshape[i])
             src_idx += 1
             dst_idx += 1
         elif newshape[i] == 0:
```
Comment (AndrewZhaoLuo) [inline python/tvm/relay/op/dyn/_transform.py:52]: won't this codepath not be hit now? this seems a little sus (since above is >= 0 now). That seems incorrect to cover the case where allowzero != 0

Comment (AndrewZhaoLuo) [reply python/tvm/relay/op/dyn/_transform.py:52]: Isn't > 0 the comparison we want?

File: tests/python/contrib/test_cmsisnn/utils.py [hunk 8]
```diff
@@ -290,7 +290,7 @@ def generate_ref_data_tflite(model):
 
 
 def create_conv2d_tflite_model(ifm_shape, kernel_shape, strides, dilation, padding, activation):
-    """ This method prepares TFlite graph with a single Conv2d layer """
```
Comment (AndrewZhaoLuo) [inline tests/python/contrib/test_cmsisnn/utils.py:293]: probably undo this change to not gum up the commit log

File: python/tvm/script/tir/__init__.pyi [hunk 9]
```diff
@@ -226,6 +226,7 @@ def alloc_buffer(
 """
 special_stmt - Reads/Writes
 """
+
```
Comment (AndrewZhaoLuo) [inline python/tvm/script/tir/__init__.pyi:229]: probably undo this change to not gum up the commit log

**Comment by AndrewZhaoLuo** [comment]
Ill look closer tomorrow, for now some initial stuff

File: python/tvm/relay/op/dyn/_transform.py (hunk 7 above)
Comment (jwfromm) [reply python/tvm/relay/op/dyn/_transform.py:52]: Yes you're right, I think this is a mistake that got left over during debugging.

File: include/tvm/topi/detail/ravel_unravel.h (hunk 5 above)
Comment (AndrewZhaoLuo) [reply include/tvm/topi/detail/ravel_unravel.h]: this change is gone now right?

**Review by AndrewZhaoLuo — COMMENTED** [review]
Nice PR, and easy to read. Looks like you have some test failures though. I'll approve once tests pass

**Review by jwfromm — APPROVED** [review]

**Comment by jwfromm** [comment]
Thanks for making this change @CircleSpin! The PR is now merged.
